### PR TITLE
Feature/aggregations

### LINKED ISSUE #6: Add filtering to EMQ-Dashboard
### Definition of done
- Filtering by
  - [x] Date range
  - [ ] Message type
  - [ ] QoS
  - [ ] Topic

### Reference
- [Available analytics data](https://github.com/apinf/emq-analytics-dashboard/blob/master/dashboard/DATA.md)

### PR DESCRIPTION
**Pull request by frenchbread**
### Changes
- Update packages
- Added eslint & babel
- Completely refactored to support elasticsearch aggregations (all parsing is now done on server-side, client-side only renders & watches for filter updates)
- Added filtering by:
  - date range
  - log type
  - granularity
- Added loading/request metrics to https://github.com/apinf/docs/tree/master/docs/develop/emq

### How it looks

screen shot 2017-03-16 at 14 55 15

### How fast it loads

- Very fast
- Instantly

### TODOs

- [x] Show metrics for X&Y axis
- [x] Remove background grid
- [x] Hourly/daily/weekly visualization

Closes #6

### CODE REVIEW

**Comment by frenchbread** [comment]
Please 👀

**Comment by bajiat** [comment]
@Nazarah Can you comment from the UX point of view?

**Comment by brylie** [comment]
@frenchbread can you give us some clearly defined metrics here? E.g. how many records is the dashboard loading? How much time (in milliseconds) do the following stages take (basically, what does the Chrome timeline contain)?

- data transfer
- ui paint

**Comment by brylie** [comment]
Also, what do the performance metrics look like with 10x records? 100x? A rough estimate is fine, so we can get an idea of how this would scale.

**Comment by frenchbread** [comment]
### Server side
The server side request to ES takes around `43-63` milliseconds.

screen shot 2017-03-07 at 13 03 31

### Client side
On client side, the total page load takes around `2-2.5` seconds, and the actual rendering around `300` milliseconds.

screen shot 2017-03-07 at 13 16 59

**Comment by brylie** [comment]
Excellent, and how many records are you retrieving? That way we can figure out the relationship between the number of records and performance time.

**Comment by frenchbread** [comment]
As of now the dataset that goes to the charts looks like this:

screen shot 2017-03-07 at 13 18 50

So it does not really matter how much items are there since they are not grouped "manually" (by dc.js) but are pre-aggregated. The only load I can think of can come from selecting the maximum possible date range, e.g. years, decades? In that case the list of records will grow. Right now, by default, around 30 days (depending on a month) are shown.

I can go ahead & populate ES with more sample data, say for last 2+ years, and compare the results.

**Comment by brylie** [comment]
Cool, yes. Keep in mind that the aggregation step takes time too.

screenshot_20170307_132924

**Comment by frenchbread** [comment]
@brylie 

> Keep in mind that the aggregation step takes time too.

This is what "server-side" metric measures.


Here is a range for all the data in the ES (1st Jan - 3rd Mar): 3069 items total. And it also takes from `40-60` ms to request data.

```
[{"key":"MQT connections over time","values":[{"x":1483308000000,"y":33},{"x":1483394400000,"y":107},{"x":1483480800000,"y":116},{"x":1483567200000,"y":46},{"x":1483653600000,"y":35},{"x":1483740000000,"y":72},{"x":1483826400000,"y":88},{"x":1483912800000,"y":68},{"x":1483999200000,"y":32},{"x":1484085600000,"y":48},{"x":1484172000000,"y":115},{"x":1484258400000,"y":47},{"x":1484344800000,"y":88},{"x":1484431200000,"y":20},{"x":1484517600000,"y":108},{"x":1484604000000,"y":41},{"x":1484690400000,"y":52},{"x":1484776800000,"y":101},{"x":1484863200000,"y":60},{"x":1484949600000,"y":96},{"x":1485036000000,"y":65},{"x":1485122400000,"y":42},{"x":1485208800000,"y":118},{"x":1485295200000,"y":106},{"x":1485381600000,"y":50},{"x":1485468000000,"y":92},{"x":1485554400000,"y":101},{"x":1485640800000,"y":93},{"x":1485727200000,"y":94},{"x":1485813600000,"y":43},{"x":1485900000000,"y":116},{"x":1485986400000,"y":24},{"x":1486072800000,"y":105},{"x":1486159200000,"y":80},{"x":1486245600000,"y":29},{"x":1486332000000,"y":72},{"x":1486418400000,"y":52},{"x":1486504800000,"y":116},{"x":1486591200000,"y":40},{"x":1486677600000,"y":67},{"x":1486764000000,"y":27},{"x":1486850400000,"y":86},{"x":1486936800000,"y":67},{"x":1487023200000,"y":0},{"x":1487109600000,"y":0},{"x":1487196000000,"y":0},{"x":1487282400000,"y":0},{"x":1487368800000,"y":0},{"x":1487455200000,"y":0},{"x":1487541600000,"y":0},{"x":1487628000000,"y":0},{"x":1487714400000,"y":0},{"x":1487800800000,"y":0},{"x":1487887200000,"y":0},{"x":1487973600000,"y":0},{"x":1488060000000,"y":0},{"x":1488146400000,"y":0},{"x":1488232800000,"y":0},{"x":1488319200000,"y":0},{"x":1488405600000,"y":0},{"x":1488492000000,"y":11}]}]
```

**Comment by brylie** [comment]
Here is a revised diagram:

screenshot_20170307_133407

From your understanding, what is the point I am trying to make here?

**Comment by brylie** [comment]
Lets conduct a couple more **server-side aggregation** experiments, as mentioned above:

- 3,000 - 5,000 records
- 10,000 records
- 50,000 records
- 100,000

That way we can get more insight into how the *aggregation* step might vary as a function of the number of records.

**Comment by frenchbread** [comment]
> From your understanding, what is the point I am trying to make here?

@brylie Isn't it "measuring time aggregation takes"? Since aggregation is done on elasticsearch side, search execution metric is exposed in the response object (`3` milliseconds in this case):

screen shot 2017-03-07 at 14 47 15

**Comment by frenchbread** [comment]
### Update

I've modified the server code and moved ElasticSearch client initialisation out of the Meteor method,  this way it saves around `20` milliseconds more.

**Comment by brylie** [comment]
> Since aggregation is done on elasticsearch side, search execution metric is exposed in the response object

Thanks for the example. 3ms aggregation with 3,000 records is quite fast. Can we try a couple of experiments with larger data, say 10,000 or 50,000 record aggregation?

**Comment by frenchbread** [comment]
@brylie I'm on it. 🚀

**Comment by frenchbread** [comment]
Generated sample data & tested request/aggregation time, collected metrics & saved to `docs/request-aggregation-metrics.csv`

### Results

As seen in the CSV file, growing amount of analytics data for **fixed amount of days** (in this case 66 days ~ 2 months) does not slow down request/aggregation time. In fact, referring to numbers speed is even increasing:

| Items | Server -> ES request (ms) | ES search execution (ms) |
| ------ | ----------------------------------- | -------------------- |
| 13k | 42.9 | 3.9 |
| 20k | 30 | 3.7 |
| 50k | 28.7 | 2.3 |

**Comment by frenchbread** [comment]
As of now, for this PR provided metrics is enough I suppose. I've created related issue #8 for testing within wider date ranges (larger amount of days).

**Comment by Nazarah** [comment]
@frenchbread : great work done. I am really liking the simple look of chart here.
A few improvement suggestions:
* showing the metrics that we are meazuring in X-axis and Y-Axis along with units.
* It would have been great to remove the background grid from the graph so only the plot is seen. But if the graphs are built-in by default that way, we can skip it.

A few queries
* Are we considering any hourly/daily/weekly visualization of data here?
* From EMQTT schema, what metrics we wold be considering to show in visualization as part of API usage monitoring?
* Will we be considering Hourly/Daily/Weekly usage visualizations?

To me the client side monitoring visualization graph is looking more attention grabbing and interesting
https://cloud.githubusercontent.com/assets/2122679/23654223/5d310890-0338-11e7-8aec-5ba59a9ca50e.png
Can we somehow incorporate this sort of visualizations with EMQTT?

**Comment by frenchbread** [comment]
@Nazarah Thanks for you comments. I've updated def. of done.

> Are we considering any hourly/daily/weekly visualization of data here?

Yes, it is on the way.

> From EMQTT schema, what metrics we wold be considering to show in visualization as part of API usage monitoring?

That is still not decided. 

> Can we somehow incorporate this sort of visualizations with EMQTT?

That's good idea. What exactly are you referring to?

We could actually divide the current (general overview) chart that we have into multiple smaller ones where each chart would represent usage for specific log type (e.g. `on_client_connected`, `on_client_subscribed` etc.) and color them differently.

How does that sound? @bajiat  @brylie @Nazarah

**Comment by brylie** [comment]
> From EMQTT schema, what metrics we wold be considering to show in visualization as part of API usage monitoring?

@frenchbread can you give us some examples of the types of metrics we might expect? E.g. the name of one or more metric(s) and the data type(s). That way we can at least start sketching some ideas.

**Comment by frenchbread** [comment]
Here they are https://github.com/apinf/emq-analytics-dashboard/blob/master/dashboard/DATA.md#type-string-possible-options

**Comment by frenchbread** [comment]
Added granularity filter:

#### By month

screen shot 2017-03-09 at 14 33 51

#### By week

screen shot 2017-03-09 at 14 33 57

#### By day

screen shot 2017-03-09 at 14 34 06

#### By hour

screen shot 2017-03-09 at 14 34 13

**Comment by frenchbread** [comment]
This is what I meant by 

> We could actually devide the current (general overview) chart that we have into multiple smaller ones where each chart would represent usage for specific log type (e.g. on_client_connected, on_client_subscribed etc.) and color them differently.

screen shot 2017-03-09 at 16 40 55

Are we going to keep this (& I commit changes here)?

**Comment by frenchbread** [comment]
There are some drawbacks here: 
- When selecting a wide dates range (e.g. 2 years) & granularity `hour` - dashboard freezes

In other cases works fine

**Comment by brylie** [comment]
Why does the dashboard freeze with large date ranges?

**Comment by frenchbread** [comment]
I assume it happens due rendering data for large amount of days (> 600 days).

**Comment by frenchbread** [comment]
`Idea`: we can create use cases for filter queries when dashboard freezes, hide chart and show placeholder message e.g. "Too much data to show" or "Select smaller dates range"..

**Comment by brylie** [comment]
> I assume it happens due rendering data for large amount of days (> 600 days).

How can you test that hypothesis?

**Comment by brylie** [comment]
Another idea is to have 'default' granularity settings. E.g. when selecting two years span, defaulting the query to monthly (or daily).

**Comment by frenchbread** [comment]
Removed grid lines & changed datepicker to `pikaday` library.

### TODO:
- Move `log type` filter to drop down menu

**Comment by brylie** [comment]
Is this ready for review, or will it be under construction a bit longer?

**Comment by frenchbread** [comment]
@brylie This is not yet ready.

**Comment by frenchbread** [comment]
Moved filtering by message type to its own select-picker.

screen shot 2017-03-14 at 16 26 33

Please review.

**Comment by brylie** [comment]
Woohoo! 😀

**Comment by frenchbread** [comment]
@brylie Could you merge this? If you are going to test it, I can send you ES host with sample data.

File: core/lib/client/pikaday.css [hunk 1]
```diff
@@ -0,0 +1,222 @@
+@charset "UTF-8";
```
Comment (brylie) [inline core/lib/client/pikaday.css]: Is it difficult/impossible to import this CSS from the NPM package?

File: dashboard/client/index.js [hunk 2]
```diff
@@ -1,137 +1,146 @@
...
-        {
-          date: {
-            order: 'desc',
-          },
-        },
-      ],
-    },
-  };
+  instance.updateQuery = () => {
+    const from = FlowRouter.getQueryParam('from') || moment().subtract(1, 'month').format('YYYY-MM-DD')
```
Comment (brylie) [inline dashboard/client/index.js:94]: Why do we duplicate this here? The FlowRouter `from`, `to`, `granularity`, and `type` query parameters should be set onRender, correct?

Comment (brylie) [reply dashboard/client/index.js:94]: Is there a case where this code is failing without the `||`?

File: dashboard/client/index.js [hunk 3]
```diff
@@ -1,137 +1,146 @@
...
+    const to = FlowRouter.getQueryParam('to') || moment().format('YYYY-MM-DD')
+    const granularity = FlowRouter.getQueryParam('granularity') || 'day'
+    const type = FlowRouter.getQueryParam('type') || ''
+
+    // For metics collection
+    // const a = moment(from, 'YYYY-MM-DD')
```
Comment (brylie) [inline dashboard/client/index.js]: Clean up commented code.

File: dashboard/client/index.js [hunk 4]
```diff
@@ -1,137 +1,146 @@
...
+    // const b = moment(to, 'YYYY-MM-DD')
+    // console.log(`${b.diff(a, 'days')} days shown.`)
+
+    const q = instance.opts.body.query.bool.must
```
Comment (brylie) [inline dashboard/client/index.js]: What does `q` refer to? Lets give this variable a semantic name, so we can [avoid mental mapping](https://github.com/ryanmcdermott/clean-code-javascript#avoid-mental-mapping).

File: dashboard/server/methods.js [hunk 5]
```diff
@@ -1,18 +1,84 @@
...
+  'on_session_terminated',
+  'on_message_publish',
+  'on_message_delivered',
+  'on_message_acked'
+]
 
 Meteor.methods({
-  getEsData (opts) {
+  getAggr (opts) {
+    // const searches = _.map(types, type => {
```
Comment (brylie) [inline dashboard/server/methods.js]: Clean up commented code.

File: dashboard/server/methods.js [hunk 6]
```diff
@@ -1,18 +1,84 @@
...
+    // })
+    //
+    // return Promise.all(searches)
+    //   .then(res => res)
+    //   .catch(err => { throw new Meteor.Error(err.message) })
+
+    const startTime = new Date().getTime()
+    return client.search(opts).then(res => {
+      const took1 = new Date().getTime() - startTime
+      console.log(`--------------`)
```
Comment (brylie) [inline dashboard/server/methods.js]: Clean up console logs.

File: dashboard/server/methods.js [hunk 7]
```diff
@@ -1,18 +1,84 @@
...
+        x: moment(item.key_as_string, 'DD-MM-YYYY').valueOf(),
+        y: item.doc_count
+      }))
 
-    return client.search(opts).then((res) => {
-      return res;
-    }, (err) => {
-      throw new Meteor.Error(err.message);
-    });
+      return [
```
Comment (brylie) [inline dashboard/server/methods.js]: This function name `getAggr` basically says 'get aggregate data', but we are adding chart style parameters here. [Functions should do one thing](https://github.com/ryanmcdermott/clean-code-javascript#functions-should-do-one-thing), so this function should only return the aggregate data without any chart style rules.

File: dashboard/server/methods.js [hunk 8]
```diff
@@ -1,18 +1,84 @@
...
+  'on_session_unsubscribed',
+  'on_session_terminated',
+  'on_message_publish',
+  'on_message_delivered',
+  'on_message_acked'
+]
 
 Meteor.methods({
-  getEsData (opts) {
+  getAggr (opts) {
```
Comment (brylie) [inline dashboard/server/methods.js]: While this function name pretty much describes what it does, it should also be [meaningful and pronouncable](https://github.com/ryanmcdermott/clean-code-javascript#use-meaningful-and-pronounceable-variable-names). `getAggr` is *almost* meaningful and *sort-of* pronouncable, but would be much better as `getAggregateData`.

File: dashboard/server/startup.js [hunk 9]
```diff
@@ -3,35 +3,86 @@
 // import moment from 'moment'
```
Comment (brylie) [inline dashboard/server/startup.js]: Clean up commented code.

File: dashboard/server/startup.js [hunk 10]
```diff
@@ -3,35 +3,86 @@
...
 // })
 //
-// const dateStart = moment(new Date("January 01, 2017 10:10:00"))
-// const dateNow = moment()
+// const dateStart = moment(new Date('April 7, 2017 10:10:00'))
+// const dateNow = moment(new Date('September 5, 2017 10:10:00'))
 //
 // const diff = dateNow.diff(dateStart, 'days')
 //
+// const types = [
```
Comment (brylie) [inline dashboard/server/startup.js]: Clean up commented code.

File: dashboard/server/startup.js [hunk 11]
```diff
@@ -3,35 +3,86 @@
...
 //       id: uuid.v4(),
 //       body: {
 //         date: logDate.toDate(),
-//         type: 'on_message_delivered',
+//         type: randomType,
 //         username: 'admin'
 //       }
 //     }
+//
+//     if (randomType === 'on_client_subscribe') {
```
Comment (brylie) [inline dashboard/server/startup.js]: Clean up commented code.

File: docs/request-aggregation-metrics.csv [hunk 12]
```diff
@@ -0,0 +1,33 @@
+Items count,Days amount,Server → ES requrest (ms),ES search exec (ms)
```
Comment (brylie) [inline docs/request-aggregation-metrics.csv]: Consider moving this to the `apinf/docs` repository, to a folder containing research and testing related to your EMQ work.

**Review by brylie — CHANGES_REQUESTED** [review]

**Comment by brylie** [comment]
This looks really good from a UI perspective! I made some suggestions on how to improve the code style, so our code conforms to our [developer expectations](https://github.com/apinf/platform/blob/develop/CONTRIBUTING.md#code-quality).

**Comment by phanimahesh** [comment]
Was this checked with live ES data? If not, @frenchbread let's do a quick check with live data once. First thing tomorrow morning, unless you are occupied. We exchanged docs on event format but still.

File: dashboard/client/index.js (hunk 2 above)
Comment (frenchbread) [reply dashboard/client/index.js:94]: There is a case when URL does not contain `from` or `to` params, but we still need to have default date range to request data for.

**Comment by frenchbread** [comment]
Added suggested changes. Please review

**Comment by frenchbread** [comment]
After new elasticsearch plugin for emq is deployed, it would require to refactor a little bit in order to fit updated analytics logs schema. But this could be done in separate PR

**Comment by frenchbread** [comment]
@phanimahesh As we discussed in the chat, yes, dashboard has been tested with live/real ES data. For purposes of testing (with wider date-ranges & collecting metrics of loading/rendering) some dummy data was added in addition to 'real' generated by the plugin.

File: dashboard/client/index.js [hunk 13]
```diff
@@ -41,8 +41,18 @@ Template.dashboard.onCreated(function () {
     }
   }
 
-  instance.getAggrData = opts => new Promise((resolve, reject) => {
-    Meteor.call('getAggr', opts, (err, res) => {
+  instance.getChartData = opts => new Promise((resolve, reject) => {
+    Meteor.call('getAggregatedData', opts, (err, res) => {
+      if (err) reject(err)
+      Meteor.call('parseDataForNvd', res, (err1, res1) => {
```
Comment (brylie) [inline dashboard/client/index.js]: Should this be a Meteor method, or just a client side funciton? I am asking, be cause Meteor methods are mainly server-side (with client side stub). That means, in this scenario, we are getting aggregated data from the server and then sending it back to the server to be parsed and returned again.

File: dashboard/client/index.js [hunk 14]
```diff
@@ -41,8 +41,18 @@ Template.dashboard.onCreated(function () {
...
+        if (err1) reject(err1)
+        resolve(res1)
+      })
+    })
+  })
+
+  instance.parseDataForNvd = data => new Promise((resolve, reject) => {
+    Meteor.call('parseDataForNvd', data, (err, res) => {
```
Comment (brylie) [inline dashboard/client/index.js]: Again, it seems like this parsing step (preparing it for NVD3 should either:

- happen as a sub-function call in the `getChartData` server method
- happen as a client side parsing function

It seems inefficient that we would get data from the server and send it back to the server for further parsing.

File: dashboard/client/index.js [hunk 15]
```diff
@@ -106,32 +116,27 @@ Template.dashboard.onCreated(function () {
...
-    // console.log(`${b.diff(a, 'days')} days shown.`)
-
-    const q = instance.opts.body.query.bool.must
+    const mustQuery = instance.opts.body.query.bool.must
 
-    const range = _.find(q, obj => typeof obj.range !== 'undefined')
+    const range = _.find(mustQuery, obj => typeof obj.range !== 'undefined')
     if (range) {
-      _.remove(q, _.find(q, obj => typeof obj.range !== 'undefined'))
+      _.remove(mustQuery, _.find(mustQuery, obj => typeof obj.range !== 'undefined'))
```
Comment (brylie) [inline dashboard/client/index.js]: This line of code deserves a human friendly comment.

File: dashboard/client/index.js [hunk 16]
```diff
@@ -106,32 +116,27 @@ Template.dashboard.onCreated(function () {
...
-    // For metics collection
-    // const a = moment(from, 'YYYY-MM-DD')
-    // const b = moment(to, 'YYYY-MM-DD')
-    // console.log(`${b.diff(a, 'days')} days shown.`)
-
-    const q = instance.opts.body.query.bool.must
+    const mustQuery = instance.opts.body.query.bool.must
 
-    const range = _.find(q, obj => typeof obj.range !== 'undefined')
+    const range = _.find(mustQuery, obj => typeof obj.range !== 'undefined')
```
Comment (brylie) [inline dashboard/client/index.js]: Add comment here describing this line of code in plain language.

File: dashboard/client/index.js [hunk 17]
```diff
@@ -106,32 +116,27 @@ Template.dashboard.onCreated(function () {
...
     if (range) {
-      _.remove(q, _.find(q, obj => typeof obj.range !== 'undefined'))
+      _.remove(mustQuery, _.find(mustQuery, obj => typeof obj.range !== 'undefined'))
     }
 
-    const match = _.find(q, obj => typeof obj.match !== 'undefined')
+    const match = _.find(mustQuery, obj => typeof obj.match !== 'undefined')
```
Comment (brylie) [inline dashboard/client/index.js]: Add comment.

File: dashboard/client/index.js [hunk 18]
```diff
@@ -106,32 +116,27 @@ Template.dashboard.onCreated(function () {
...
     if (match) {
-      _.remove(q, _.find(q, obj => typeof obj.match !== 'undefined'))
+      _.remove(mustQuery, _.find(mustQuery, obj => typeof obj.match !== 'undefined'))
```
Comment (brylie) [inline dashboard/client/index.js]: Add comment.

**Review by brylie — CHANGES_REQUESTED** [review]

File: dashboard/client/index.js (hunk 13 above)
Comment (frenchbread) [reply dashboard/client/index.js]: This is a server-side method created to decrease the load on client-side (browser).

> That means, in this scenario, we are getting aggregated data from the server and then sending it back to the server to be parsed and returned again.

Right.

Comment (brylie) [reply dashboard/client/index.js]: I think we should **reduce the number of round trips**. If we want to decrease the client load by performing a second aggregation on the server, then **send the data only after it has been aggregated properly**.

**Comment by frenchbread** [comment]
Added comments

**Comment by frenchbread** [comment]
Removed second meteor method. Please :mag_right:

**Pull request merged**